### PR TITLE
Feat: support terragrunt

### PR DESCRIPTION
**Pull request by TomerHeber**
I may not be familiar with all the edge cases and caveats. Hopefully, I haven't missed anything.
I would like to add acceptance tests as a follow-up PR.

#146

### CODE REVIEW

File: cli/args.go [hunk 1]
```diff
@@ -29,6 +30,7 @@ func InitArgs() (Args, bool) {
 	fs.StringVar(&args.Filter, "filter", ".*", "Only apply tags to the selected resource types (regex)")
 	fs.BoolVar(&args.Verbose, "verbose", false, "Enable verbose logging")
 	fs.BoolVar(&args.Rename, "rename", true, "Keep the original filename or replace it with <basename>.terratag.tf")
+	fs.BoolVar(&args.Terragrunt, "terragrunt", false, "Tags all the terraform files under .terragrunt-cache")
```
Comment (TomerHeber) [inline cli/args.go:33]: Added terragrunt support as a flag.

File: cmd/terratag/main.go [hunk 2]
```diff
@@ -17,7 +17,9 @@ func main() {
 	}
 	initLogFiltering(args.Verbose)
 
-	terratag.Terratag(args)
+	if err := terratag.Terratag(args); err != nil {
```
Comment (TomerHeber) [inline cmd/terratag/main.go:20]: unrelated, but missed printing out errors.

File: internal/terraform/terraform.go [hunk 3]
```diff
@@ -73,21 +74,63 @@ func GetResourceType(resource hclwrite.Block) string {
...
+	rootDir += "/.terragrunt-cache"
+
+	var tfFiles []string
+	if err := filepath.WalkDir(rootDir, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			log.Printf("[WARN] skipping %s due to an error: %v", path, err)
+			return filepath.SkipDir
+		}
+
+		if strings.HasSuffix(path, ".tf") {
```
Comment (TomerHeber) [inline internal/terraform/terraform.go:121]: assuming anything with *.tf needs to be processed.

File: internal/tfschema/tfschema.go [hunk 4]
```diff
@@ -84,7 +86,33 @@ func detectProviderName(resource hclwrite.Block) (string, error) {
 	return extractProviderNameFromResourceType(terraform.GetResourceType(resource))
 }
 
-func getClient(providerName string, dir string) (tfschema.Client, error) {
+func getTerragruntPluginPath(dir string) string {
```
Comment (TomerHeber) [inline internal/tfschema/tfschema.go:89]: It's non trivial to locate the "plugin/client" path to .terraform (since the folder is a bunch of randomly generated strings). Therefore, had to locate it.

E.g.:  ./.terragrunt-cache/yHtqnMrVQOISIYxobafVvZbAAyU/ThyYwttwki6d6AS3aD5OwoyqIWA/.terraform

File: cli/args.go (hunk 1 above)
Comment (yaronya) [inline cli/args.go:33]: Why do we need this to be optional?
I mean, couldn't we simply identify if there's a cache folder - and if there is, tag its content?

One reason that I thought of - is to alert the user that he needs to run `init` before using `terratag` in the use case of `terragrunt`.
If that was your thought - I'd suggest to introduce a more generic `type` flag that defaults to `terraform` so we can later even support other IaC types

File: internal/terraform/terraform.go [hunk 5]
```diff
@@ -73,21 +74,63 @@ func GetResourceType(resource hclwrite.Block) string {
...
+func GetFilePaths(dir string, terragrunt bool) ([]string, error) {
+	if terragrunt {
+		return getTerragruntFilePath(dir)
+	} else {
+		return getTerraformFilePaths(dir)
+	}
+}
+
+func getTerragruntFilePath(rootDir string) ([]string, error) {
+	rootDir += "/.terragrunt-cache"
```
Comment (yaronya) [inline internal/terraform/terraform.go:112]: Can we DRY this constant a bit? i.e using some common func of `getRootDir` that receives an IaC type

File: internal/terraform/terraform.go [hunk 6]
```diff
@@ -73,21 +74,63 @@ func GetResourceType(resource hclwrite.Block) string {
...
+
+	var tfFiles []string
+	if err := filepath.WalkDir(rootDir, func(path string, d fs.DirEntry, err error) error {
```
Comment (yaronya) [inline internal/terraform/terraform.go:115]: Consider avoid using inline syntax for this function declaration. I find it a little confusing, especially for newcomers to golang

File: terratag.go [hunk 7]
```diff
@@ -45,24 +45,24 @@ func Terratag(args cli.Args) error {
...
+	counters := tagDirectoryResources(args.Dir, args.Filter, matches, args.Tags, args.IsSkipTerratagFiles, *tfVersion, args.Rename, args.Terragrunt)
 	log.Print("[INFO] Summary:")
 	log.Print("[INFO] Tagged ", counters.taggedResources, " resource/s (out of ", counters.totalResources, " resource/s processed)")
 	log.Print("[INFO] In ", counters.taggedFiles, " file/s (out of ", counters.totalFiles, " file/s processed)")
 
 	return nil
 }
 
-func tagDirectoryResources(dir string, filter string, matches []string, tags string, isSkipTerratagFiles bool, tfVersion convert.Version, rename bool) counters {
+func tagDirectoryResources(dir string, filter string, matches []string, tags string, isSkipTerratagFiles bool, tfVersion convert.Version, rename bool, terragrunt bool) counters {
```
Comment (yaronya) [inline terratag.go:65]: Can we move away to an `args` struct/object and not use so many parameters at once? 🙏

File: terratag.go [hunk 8]
```diff
@@ -99,7 +99,7 @@ func tagDirectoryResources(dir string, filter string, matches []string, tags str
 	return total
 }
 
-func tagFileResources(path string, dir string, filter string, tags string, tfVersion convert.Version, rename bool) (*counters, error) {
+func tagFileResources(path string, dir string, filter string, tags string, tfVersion convert.Version, rename bool, terragrunt bool) (*counters, error) {
```
Comment (yaronya) [inline terratag.go:102]: Same comment as above 🙏

**Review by yaronya — COMMENTED** [review]
Nicely done
Left some comments/suggestions

File: cli/args.go (hunk 1 above)
Comment (TomerHeber) [reply cli/args.go:33]: Yes, that was my thought.
I will modify it to "type".

File: internal/terraform/terraform.go (hunk 5 above)
Comment (TomerHeber) [reply internal/terraform/terraform.go:112]: Yes!

File: internal/terraform/terraform.go (hunk 6 above)
Comment (TomerHeber) [reply internal/terraform/terraform.go:115]: It's actually done in this way on purpose (all examples are like this) and is standard in go. The main reason is that I need to use the variables in lines 112 and 114.
I could move it to a function that returns a function (to store the variable's context) if you think it will make it more readable. Let me know what you prefer.

File: terratag.go (hunk 7 above)
Comment (TomerHeber) [reply terratag.go:65]: Sure!

File: terratag.go (hunk 8 above)
Comment (TomerHeber) [reply terratag.go:102]: Sure!

File: internal/terraform/terraform.go (hunk 6 above)
Comment (yaronya) [reply internal/terraform/terraform.go:115]: This is no biggie so if you feel it's a valid standard then let's leave it be 👍

File: terratag.go (hunk 7 above)
Comment (TomerHeber) [reply terratag.go]: refactored.

File: internal/terraform/terraform.go (hunk 5 above)
Comment (TomerHeber) [reply internal/terraform/terraform.go]: refactored.

**Comment by TomerHeber** [comment]
For follow-up PR:
@yaronya - 
1. happy to add documentation - some guidance would be great. (README.md?)
2. add acceptance test (need a valid HCL to use, do you guys have something?).

Thanks!

**Comment by yaronya** [comment]
> For follow-up PR:
> 
> @yaronya - 
> 
> 1. happy to add documentation - some guidance would be great. (README.md?)
> 
> 2. add acceptance test (need a valid HCL to use, do you guys have something?).
> 
> 
> 
> Thanks!

@TomerHeber you can have a look at the `env0/templates` repository and see if you can find a decent example there. Feel free to reach out if you need any further guidance 

Regarding the docs - let's open a PR first and take it there together

**Review by yaronya — APPROVED** [review]

**Comment by TomerHeber** [comment]
> > For follow-up PR:
> > @yaronya -
> > 
> > 1. happy to add documentation - some guidance would be great. (README.md?)
> > 2. add acceptance test (need a valid HCL to use, do you guys have something?).
> > 
> > Thanks!
> 
> @TomerHeber you can have a look at the `env0/templates` repository and see if you can find a decent example there. Feel free to reach out if you need any further guidance
> 
> Regarding the docs - let's open a PR first and take it there together

@yaronya - sounds good. 
I'll work on adding tests once I'm back (in a week).
I'll follow-up on docs once I'm back.

I'll push this in and keep the issue open.
